### PR TITLE
Feature/justiceotuya/signup

### PR DESCRIPTION
**Pull request by justiceotuya**
finished the signup page

### CODE REVIEW

File: components/Authentication/components/Login.jsx [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    Input,
+    Button,
+    Typography
+} from 'antd';
+import 'antd/dist/antd.css';
+import './Authentication.css';
+import Router from 'next/router';
+import LoginImage from '../../../static/login.svg';
+import {
+    loginUsernameInputError,
```
Comment (EngrEric) [inline components/Authentication/components/Login.jsx]: constants should be IN upper-case and separated with underscore. that is
`import {
    USERNAME_INPUT_ERROR,    PASSWORD_INPUT_ERROR
} from '../constants';`

Comment (justiceotuya) [reply components/Authentication/components/Login.jsx]: done

File: components/Authentication/components/Login.jsx [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                        </Form.Item>
+
+                        <Form.Item>
+                            <a className="login-form-forgot" href="/password_reset">Forgot password</a>
+                            <Button type="primary" htmlType="submit" className="login-form-button" loading={loading} onClick={this.enterLoading}>
+                                Log in
+                            </Button>
+                            Or
+                            <a className="login-form-register" href="/signup"> register now!</a>
+                        </Form.Item>
```
Comment (EngrEric) [inline components/Authentication/components/Login.jsx:81]: @justiceotuya  create a function that returns your `<Form.Item>...</Form.Item>` 

```
function generateLoginForms(decorator) {
   return INPUTS.map(input => {
        const { field, iconType, placeholder, rules, type } = input;
       return (
            <FormItem
               key={field}
            >
                {
                    decorator(field, {
                       rules,
                    })(
                        <Input
                            placeholder={placeholder}
                            prefix={<Icon type={iconType} />}
                            type={type}
                        />
                   )
                }
            </FormItem>
        );
    });
}

THE VALUE OF `INPUT` WILL LOOK LIKE

INPUTS = [
    {
        field: '',
        iconType: "",
        placeholder: "",
        rules: [{ message: "", required: true }],
    },
   {
        field: '',
        iconType: "",
       placeholder: "",
       type: "",
    },
];

THEN YOUR `<form></form>` COMPONENT WILL LOOK LIKE

                <Form>
                    {generateLoginForms(getFieldDecorator)}
                </Form>

File: components/Authentication/components/SignUp.jsx [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    Checkbox,
+    Button,
+    AutoComplete,
+    Typography
+} from 'antd';
+import 'antd/dist/antd.css';
+import './Authentication.css';
+import Router from 'next/router';
+import RegistrationImage from '../../../static/register.svg';
+import {
```
Comment (EngrEric) [inline components/Authentication/components/SignUp.jsx:9]: all constants should be uppercase and separated with underscore as said above

File: components/Authentication/components/SignUp.jsx [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        const { value } = e.target;
+        this.setState({ confirmDirty: this.state.confirmDirty || !!value });
+        console.log('dirty state', this.state.confirmDirty);
+        console.log('value', !!value);
+    }
+
+    compareToFirstPassword = (rule, value, callback) => {
+        const { form } = this.props;
+        if (value && value !== form.getFieldValue('password')) {
+            callback('The Two passwords that you enter is inconsistent!');
```
Comment (EngrEric) [inline components/Authentication/components/SignUp.jsx:83]: move all strings to constants

File: components/Authentication/components/SignUp.jsx [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                    Router.push('/timeline');
+                }, 1000);
+            }
+        });
+    }
+
+    handleConfirmBlur = e => {
+        const { value } = e.target;
+        this.setState({ confirmDirty: this.state.confirmDirty || !!value });
+        console.log('dirty state', this.state.confirmDirty);
```
Comment (EngrEric) [inline components/Authentication/components/SignUp.jsx]: remove all `console.log()`
you want to use it when your PR is tagged WIP (work in progress).

File: components/Authentication/components/SignUp.jsx [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                        </Form.Item>
+
+                        <Form.Item
+                          label="Password"
+                        >
+                            {getFieldDecorator('password', {
+                                rules: [{
+                                    required: true, message: signupPasswordInputError,
+                                }, {
+                                    validator: this.validateToNextPassword,
```
Comment (EngrEric) [inline components/Authentication/components/SignUp.jsx]: Do the same thing here as said in Loginform component

File: components/Authentication/constants.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+const signupPasswordInputError = 'Please input your password!';
+const signupConfirmPasswordInputError = 'Please confirm your password!';
+
+export {
+    loginUsernameInputError,
+    loginPasswordInputError,
+    signupNameInputError,
+    signupEmailInputError,
+    signupValidEmailError,
+    signupPasswordInputError,
```
Comment (EngrEric) [inline components/Authentication/constants.js]: all constants should be in uppercase

**Review by EngrEric — CHANGES_REQUESTED** [review]
looks good but needs some changes

**Review by Way2nnadi — DISMISSED** [review]
Looks good to me

**Review by enyeInc — DISMISSED** [review]

File: components/Authentication/components/Login.jsx [hunk 8]
```diff
@@ -29,15 +20,29 @@ class NormalLoginForm extends React.Component {
...
     handleSubmit = e => {
         const { validateFields } = this.props.form;
         e.preventDefault();
-        // validate error and simulate api response with settimeout
+
+        /**
+    *  function that is used to handle submit, This function helps to Validate the specified fields and get theirs values and errors., if the target field is not in visible area of form, form will be automatically scrolled to the target field area.
+    * @function
+    * @return {Object}  returns the values of the form
+    */
```
Comment (enyeInc) [inline components/Authentication/components/Login.jsx]: make sure you comments are indented properly

File: components/Authentication/components/LoginInputItemGenerator.jsx [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+                decorator(field, {
+                    rules,
+                })(
+                    <Input prefix={<Icon type={iconType} className={className} />} placeholder={placeholder} type={inputType} />
+                )
+            }
+        </Form.Item>
+    );
+});
+
```
Comment (enyeInc) [inline components/Authentication/components/LoginInputItemGenerator.jsx:46]: needs component propType definition

File: components/Authentication/components/SignUp.jsx [hunk 10]
```diff
@@ -12,68 +12,97 @@ import 'antd/dist/antd.css';
...
+ * @return {Object} the registration  page
  */
 class RegistrationForm extends React.Component {
     state = {
-        // confirmDirty: false,
+        confirmDirty: false,
         autoCompleteResult: [],
         iconLoading: false,
         loading: false,
+        isAgreementChecked: false,
```
Comment (enyeInc) [inline components/Authentication/components/SignUp.jsx:25]: so all booleans have to be in question format like `isAgreementChecked`.
```
isConfirmedDirty, isIconLoading, isLoading
```

**Review by enyeInc — DISMISSED** [review]
Looking real good, some minor issues to fix but code is good to be merged down

File: components/Authentication/components/SignUp.jsx [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+    */
+    handleSubmit = e => {
+        e.preventDefault();
+        const { validateFieldsAndScroll } = this.props.form;
+        const { isAgreementChecked } = this.state;
+        /**
+    *  function that is used to handle submit, This function helps to Validate the specified fields and get theirs values and errors., if the target field is not in visible area of form, form will be automatically scrolled to the target field area.
+    * @function
+    * @return {Object}  returns the values of the form
+    */
```
Comment (enyeInc) [inline components/Authentication/components/SignUp.jsx]: fix doc indent

File: components/Authentication/components/LoginInputItemGenerator.jsx [hunk 12]
```diff
@@ -8,20 +8,33 @@ import {
...
+* @param {string} id id of the input field
+* @param {string} iconType preceding icon of the input field
+* @param {function} decorator a Two-way binding for form
+* @param {string} placeholder placeholder of the input field
+* @param {Object[]} rules rules for input validation
+* @param {string} inputType type of the input element
+* @return {function} input item of the form
+*/
+
+const LOGIN_INPUT_ITEM_GENERATOR = decorator => LOGIN_INPUTS.map(input => {
```
Comment (enyeInc) [inline components/Authentication/components/LoginInputItemGenerator.jsx]: `LoginInputItemGenerator` is correct all caps and underscore separated for `constants` not components

File: components/Authentication/components/SignUp.jsx (hunk 5 above)
Comment (enyeInc) [reply components/Authentication/components/SignUp.jsx]: remove console.log on `line 70`

File: components/Authentication/components/SignUp.jsx (hunk 4 above)
Comment (enyeInc) [reply components/Authentication/components/SignUp.jsx:83]: this is still pending

File: components/Authentication/components/SignUp.jsx [hunk 13]
```diff
@@ -67,44 +60,45 @@ class RegistrationForm extends React.Component {
...
         } else {
             callback();
         }
     }
 
+    /**
+* function that is used to handle checkbox agreement
+* @function
+* @return {Boolean} controls the state of of the checkbox
+*/
```
Comment (enyeInc) [inline components/Authentication/components/SignUp.jsx:93]: fix indent of esdocs

File: components/Authentication/components/SignupInputItemGenerator.jsx [hunk 14]
```diff
@@ -1,40 +1,73 @@
...
+* @param {Object[]} rules rules for input validation
+* @param {Boolean} hasOnBlur check if the input has an onChange function attached ot it
+* @param {Boolean} hasOnChange check if the input has an onChange function attached ot it
+* @param {function} valuePropName Props of checkbox
+* @param {Boolean} isButton check if the form element is a button
+* @param {Boolean} hasFieldChildren check if the input children
+* @param {function} FieldType type of html form element
+* @return {function} input item of the form
+*/
+const SIGNUP_INPUT_GENERATOR = (actions, items, decorator) => {
```
Comment (enyeInc) [inline components/Authentication/components/SignupInputItemGenerator.jsx]: this should be `SignUpInputGenerator`

**Review by enyeInc — DISMISSED** [review]
Looks good to me, just need some minor clean up

**Review by enyeInc — DISMISSED** [review]

**Pull request merged**